src/klein/_headers.py

```diff
@@ -0,0 +1,355 @@
+# -*- test-case-name: klein.test.test_request -*-
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+HTTP headers.
+"""
+
+from attr import Factory, attrib, attrs
+from attr.validators import instance_of
+from typing import (
+    Any, AnyStr, Iterable, List, MutableSequence, Sequence, Text,
+    Tuple, Union, cast,
+)
+
+from twisted.web.http_headers import Headers
+
+from zope.interface import Attribute, Interface, implementer
+
+Any, AnyStr, Iterable, List  # Silence linter
+
+
+__all__ = ()
+
+
+# Interfaces
+
+String = Union[bytes, Text]
+
+RawHeader = Tuple[bytes, bytes]
+RawHeaders = Sequence[RawHeader]
+MutableRawHeaders = MutableSequence[RawHeader]
+
+
+class IHTTPHeaders(Interface):
+    """
+    HTTP entity headers.
+
+    HTTP headers names and values are sort-of-but-not-quite-specifically
+    expected to be text.
+
+    Because the specifications are somewhat vague, and implementations vary in
+    fidelity, both header names and values must be made available as the
+    original bytes received from the network.
+    This interface also makes them available as an ordered sequence of name and
+    value pairs so that they can be iterated in the same order as they were
+    received on the network.
+
+    The C{rawHeaders} attribute provides this as a sequence of C{(name, value)}
+    L{tuple}s.
+
+    A dictionary-like interface that maps text names to an ordered sequence of
+    text values.
+    This assumes that header name bytes are encoded as ASCII, and header value
+    bytes are encoded as ISO-8859-1.
+    """
+
+    rawHeaders = Attribute(
+        """
+        Raw header data as a tuple in the from: C{((name, value), ...)}.
+        C{name} and C{value} are bytes.
+        Headers are provided in the order that they were received.
+        Headers with multiple values are provided as separate name and value
+        pairs.
+        """
+    )  # type: RawHeaders
```

Comment:
(I did note above that if the reviewer doesn't want to review all the CI changes, to say so and I'll remove them prior to merging.  It's just hard to test everything without them.)